### PR TITLE
[TECH] Add an `emptyAllTables` method for tests

### PR DESCRIPTION
**Pull request by sroccaserra**
Add a test method to empty all data from tables.

🌍 Context
==========

We use the `DatabaseBuilder` to generate and clean data for tests, but with the limitation that it could only clean the data that it had created, it could not clean data created by other means.

🔧 Fix
======

This commit adds an `emptyAllTables` function that empties all tables, except for tooling & system tables (`knex_*` tables for instance).

This PR also changes the behaviour of tests. Before, if a `DATABASE_URL` was present, tests where still ran on sqlite. Now if a `TEST_DATABASE_URL` env variable exists, tests are run on PG.

This PR also creates a `database_migrate_and_seed` CI job, that runs migrations & seeds on a postgres database.

### CODE REVIEW

File: api/tests/test-helper.js [hunk 1]
```diff
@@ -13,7 +13,8 @@ afterEach(function() {
 
 // Knex
 const knexConfig = require('../db/knexfile');
-const knex = require('knex')(knexConfig['test']);
+const { environment } = require('../lib/settings');
+const knex = require('knex')(knexConfig[environment]);
```
Comment (sroccaserra) [inline api/tests/test-helper.js]: ⬆️  Je me pose une question ici, j'en ai besoin pour pouvoir lancer les tests sur PG mais il y avait peut-être une bonne raison (sécurité ?) d'avoir 'test' en dur, WDYT ?

File: api/tests/tooling/database-builder/database-builder.js [hunk 2]
```diff
@@ -32,4 +42,38 @@ module.exports = class DatabaseBuilder {
...
+    const tablesToDelete = _.without(tableNames,
+      'knex_migrations',
+      'knex_migrations_lock',
+      'sqlite_sequence'
+    );
+
+    for(const tableName of tablesToDelete) {
+      await this.knex.raw(this.truncateTableQuery, [tableName]);
+    }
+  }
```
Comment (Radioreve) [inline api/tests/tooling/database-builder/database-builder.js]: Le `truncate` c'est un `clear?` Parce que quand je lis truncate, je pense à troncature, donc à une séparation d'un ensemble. Le clear me parlerait plus ici. On a une fonction `clearAll`, je m'attendrait donc à voir une série de `clear` appelés les uns après les autres

File: api/tests/test-helper.js (hunk 1 above)
Comment (Radioreve) [reply api/tests/test-helper.js]: Je pense pas. D'ailleurs en locale j'ai sur une branche une fonction `testEnv()' inspirée de `developementEnv()` qui me permet également de switcher entre sqlite3 et pg. Peut-être serait-il temps de faire évoluer ce bout de code et de le commit pour que toute l'équipe puisse switcher de l'un à l'autre facilement (ou alors, temps de décommissioner sqlite3 pour de bon?)

Comment (sroccaserra) [reply api/tests/test-helper.js]: Non, merci. Du coup c'est plus raccord maintenant j'ai l'impression : avant il y avait dissonance entre `const isPG = process.env.DATABASE_URL || false` dans `seed.js` et `const knex = require('knex')(knexConfig['test']);` dans `test-helper.js`. Mais de toutes façons on n'a pas besoin de lancer les seeds sur une base de tests, donc j'ai envie de dire que les deux ont peu de conséquences si j'ai bien compris.

Comment (sroccaserra) [reply api/tests/test-helper.js]: Pour être beau, j'aurais envie que même avec `NODE_ENV='test'`, mettre une `DATABASE_URL` bascule en Postgres, ça me fait bizarre à chaque fois de mettre `NODE_ENV='production'`  pour lancer des tests sur PG.

File: api/tests/tooling/database-builder/database-builder.js (hunk 2 above)
Comment (sroccaserra) [reply api/tests/tooling/database-builder/database-builder.js]: Le `truncate table` sert à vider une table, c'est un `delete from` de toutes les lignes optimisé.

> TRUNCATE quickly removes all rows from a set of tables. It has the same effect as an unqualified DELETE on each table, but since it does not actually scan the tables it is faster. Furthermore, it reclaims disk space immediately, rather than requiring a subsequent VACUUM operation. This is most useful on large tables.

in : https://www.postgresql.org/docs/10/sql-truncate.html

Comment (jbuget) [reply api/tests/tooling/database-builder/database-builder.js]: Avec la notion de `TRUNCATE` il y a aussi la notion de "on conserve la valeur d'index courant de la table ou de la séquence". Ex : si j'ai 12 patates, quand je fais `truncate table patates`, alors je n'ai plus de lignes patates, mais à la prochaine insertion, l'ID de la nouvelle patate sera 13.

Pas de préférence qu'on appelle les méthodes `truncateXxx`, `clearXxx`, `deleteXxx` ou `emptyXxx`.

File: .circleci/config.yml [hunk 3]
```diff
@@ -137,8 +137,6 @@ jobs:
           command: |
             npm run lint
             npm test
-          environment:
-            DATABASE_URL: "postgres://circleci@localhost:5432/circleci"
```
Comment (sroccaserra) [inline .circleci/config.yml]: Est-ce qu'on a besoin d'une instance de PG ici ?
Ajouter plutôt des tests sur PG, qui testent les migrations et les seeds par exemple

**Comment by sroccaserra** [comment]
> Avant de merger, je pense qu'il faudrait faire le truc que propose @sroccaserra pour ne pas avoir à mettre `production` quand on veut juste une base `PostgreSQL`.

Done!

File: .circleci/config.yml (hunk 3 above)
Comment (sroccaserra) [reply .circleci/config.yml]: Note que les migrations fonctionnent sous PG, mais pas les seeds à ce jour.

File: api/tests/test-helper.js (hunk 1 above)
Comment (sroccaserra) [reply api/tests/test-helper.js]: Ok, done.

**Comment by sroccaserra** [comment]
Le truc qui peut coincer avec ce changement :
- si je suis en train de travailler sur `cacax`, avec un tunnel sur la prod et une variable `DATABASE_URL` qui pointe dessus, _si je lance les tests je tue la base de prod._

**Comment by jonathanperret** [comment]
Hmm. Je ne comprends pas pourquoi on pointerait `DATABASE_URL` sur la base de prod. C’est difficile de le faire accidentellement non ?
L’URL contient plus que l’adresse du serveur, il y a aussi typiquement le login/mot de passe. C’est quand même bien pratique d’avoir tout au même endroit.

**Comment by sroccaserra** [comment]
Oui, effectivement le risque est faible. J'y ai pensé quand on a ouvert un tunnel vers la prod avec @jbuget pour exécuter un script one shot. Il faudrait que le tunnel reste ouvert et qu'en plus la variable d'env `DATABASE_URL` soit exportée, ce qui n'arrive pas en pratique je crois car on n'en a pas besoin pour les scripts one shot. Je propose de traiter ce sujet dans une autre PR si c'est pertinent.
